### ISSUE TITLE
🔧 Training | Fine-Tune Phi-3-Poly with LoRA-Adapters + JSON Constraint

### ISSUE THREAD

**Issue by collinbentley1** [medlock-ai/medlock#29]
### Target Checkpoint
`weights/phi3_poly_lora.pt`

### Training Specs
| Item | Value |
|------|-------|
| Base | phi3_poly.onnx / converted to PyTorch for training |
| LoRA rank | 8 |
| LoRA α | 16 |
| Dropout | 0.05 |
| Optimiser | AdamW (β1 0.9 β2 0.95) |
| LR | 2e-4, cosine decay with 1 K warm-up steps |
| Batch size | 256 samples (global), micro-batch 8 |
| Precision | FP16 + dynamic loss scaling |
| Epochs | 3 |
| Gradient checkpoint | yes |
| JSON penalty | λ = 1.0 weight added to CE loss if generated token breaks JSON rule |

### Steps
1. Convert ONNX back to `transformers` model via `convert_onnx_to_hf.py` for training (weights identical).  
2. Inject LoRA adapters at `q_proj`, `k_proj`, `v_proj`, `o_proj`, and dense layers.  
3. Loss function:
   ```python
   loss = ce_loss + lambda_json * invalid_json_mask.mean()
   ```
   `invalid_json_mask` toggles when decoder currently outside valid JSON per stack parser.
4. Implement trainer with HuggingFace + PEFT, using gradient accumulation to fit single GPU.

### Validation

* Metrics (dev):
  * JSON exact-match ≥ 0.95
  * Rouge-L summarisation ≥ 0.32
  * MMLU proxy subset ≥ 0.63

### Outputs

* `phi3_poly_lora.pt` – PEFT diff weights
* `training_args.json`
* W&B run summary (`ENTITY/PROJECT`) or local `runs/`

### Acceptance

* `python eval/json_eval.py --model phi3_poly_lora` passes 95% JSON.
* README update `models/phi3_poly_lora/README.md` with training details.
